### DISCUSSION TITLE
Some trouble getting Coin3D and Qt to work with Visual Studio 2022

### DISCUSSION THREAD

**Discussion by harzner** [coin3d/coin#532] [Ideas]
Hey peeps,

okay, so here's my story: I had an idea about ten years ago, and I would like to revive it, using Coin3D. The idea is simple: A multiplatform 3D-Application that displays folders and text-files from a hard-drive - in 3D. I will share details later, but that's about it for now.

What I want to do is a simple 3D-Viewer application on Windows, later on Linux as well, optionally on Mac.

I followed the instructions, and got these things going:

Windows 11 PC
Boost installed - check
Visual Studio 2022 - check
Cygwin as well...
CMake 3.something
Coin3D - check - installed like a charm
Quarter - check - installed like a charm
Qt.... damn, what the heck???

I am having difficulties setting up a Visual Studio Project that just uses Coin3d, Quarter, and Qt... for some reason, I'm stuck... I also could use Cygwin, but really, all I need for now, is an easy set up on Windows 11 to create a simple "Hello World" like application that uses Coin3D and Quarter.

Qt is confusing me with their tons of tools, and all I really want is INCLUDE and LIB paths to put into my Visual Studio project...

Now...

1. I'm working on a Windows 11, but it's a dual-boot, so I can easily also use SuSE Leap... or any other with some time to invest...
2. I am using Visual Studio, but I can also easily use CygWin
3. My Main C++ file in Visual Studio does not find the headers / libs needed for Qt which I need to get Quarter running, more precisely, for some reason I can't get the QApplication included in my project.

So, like those nice people from Qt told me, I installed the fricking Qt Tools for VS, and then some, but my project just won't compile, because.... well, I guess I'm an Old-Fart, and I expect some header files, and some libraries or some code to build them...

What am I missing?

In short: Do you have a simple way to jump-start an application using Coin3D, Quarter and Qt to get rolling? On Windows, with a perspective for Linux... and, I can easily drop windows, and just get that thing going on SuSE Leap.

Thanks in advance!

Respectful greetings,
Harzner

**Comment by VolkerEnderlein** [comment]
Sure, that can be done straight forward with CMake.

Download the Qt online installer and install a suitable version (e.g. 6.8.0, 5.15.2, or ...) to a directory of your choice (e.g. C:\Qt).

I assume your Coin/Quarter is installed on Windows under `C:\Coin3D`.

A typical `CMakeLists.txt` for a `quarter` based project would then look as follows:

```
cmake_minimum_required(VERSION 3.0...3.28)
project(test_project)
find_package(Coin REQUIRED)
find_package(Quarter REQUIRED)
find_package(Qt6 COMPONENTS Widgets UiTools OpenGL OpenGLWidgets REQUIRED)
include_directories(${CMAKE_CURRENT_BINARY_DIR})
set(AUTOMOC ON)
set(AUTOUIC ON)
set(AUTORCC ON)
add_executable(test_executable main.cpp <other sources, forms, and ressources>)
target_link_libraries(test_executable PUBLIC Quarter::Quarter Coin::Coin Qt6::Widgets Qt6::UiTools Qt6::OpenGL Qt6::OpenGLWidgets)
```
Place the `CMakeLists.txt` file toplevel in your source directory.
On Windows the `CMake` calling sequence for your application based on `Qt6` would look as follows:

To configure the project (please replace <source_dir>and <build_dir> with paths to your source and build directories) call:
`cmake -B <build_dir> -S <source_dir> -G "Visual Studio 17 2022" -A x64 -D CMAKE_PREFIX_PATH=C:\Coin3D;C:\Qt\6.8.0\msvc2019_64`
This generates a Visual Studio solution/project that can be run in the IDE but can also be built using CMake with the following command:
`cmake --build <build_dir> --target ALL_BUILD --config Release`

On Linux the `CMake` calling sequence for your application would look as follows provided that Qt is installed via the package manager of your distribution:

To configure the project call:
`cmake -B <build_dir> -S <source_dir> -G "Unix Makefiles" -A x64 -D CMAKE_PREFIX_PATH=/home/username/Coin3D`
To build the Release configuration of the project call:
`cmake --build <build_dir> --target all --config Release`

Some dependencies might have been missing or can be omitted, but I hope the general procedure is clear from this commands. If further questions or issues arise, feel free to ask.
When in doubt how to link to a project I usually lookup the CI build commands in the `yml` files in the `.github` folder or the `appveyor.yml` file. 

Hope that helps you further.

Cheers, Volker

**Reply by harzner** [reply]
Hello Volker,

thank you so much, I will go ahead and follow your guidance.

I guess I'm just a little bit rusty, and I haven't done much development in the recent years.

Thank you and kind regards,
Harzner

**Comment by WizzerWorks** [comment]
Your 3D desktop idea sounds like the one that was demonstrated in Jurassic Park at the end of the movie when everyone was panicking. I worked at Silicon Graphics when this movie came out. I think the software was an internal project based on Performer (and SGI HW), and not Open Inventor. But I think it should be doable with Coin.

**Reply by harzner** [reply]
Hey WizzerWorks, it's not a 3D-Desktop that I'm thinking about, it's something else that I don't yet have answers for, BUT I dare ask you a couple of questions:

1 The keyboard that you and I are using is 100 years old, and there have been many attempts to replace it, but somehow it remains the primary tool for us to interact with our computers - along with a mouse
2 Most IDEs are 2-dimensional, and forcing us to work with a mouse, which takes at least one hand from that keyboard
3 Meanwhile, I'm convinced that our brains are doing most of the work. And what are our brains good at? Pattern recognition!
5 I am convinced that a good developer does not have to "read" the code, but would be capable to "understand" it from a distance.

Here's the idea: What if you had a text editor / file system explorer / code writing app at the same time in 3D?

I won't give away all my thoughts, but how about a "file explorer"  where you can "fly" though your file system, edit your sources, look at it "from a distance" and let your brain just take off from that? I had a couple of Indigos, so I know what SGI did at that time.

I'm getting older, and I may as well be totally wrong, but I tried that, more than 10 years ago, but I've never seen anyone take that approach.

I will try this one for one more time.

Rules:
Source-code writing in 3D ( I goddamn BET that you will recognise bad code from a mile away)
No mouse - all keyboard
Available on all platforms
Simple App on a USB-Stick - upload to your root direcory - Bam! You can fly!

Kind regards, 
Harzner

**Reply by WizzerWorks** [reply]
While I worked at SGI I had the chance to work for Charles Irby, who led development on the UI for Xerox Star (which Steve Jobs stole for the Macintosh). He was always encouraging innovative UI approaches. I wrote a visual 3D tool (using Open Inventor) that allowed firmware (graphics display lists) to be programmed at nodes in the scene graph, eventually being rendered for the new graphics technology that was being developed by his team at MIPs (follow-up to Nintendo 64).

This was an interactive approach to developing the firmware and allowed the HW engineers and test simulators to dynamically design the new display list language in context of the rendering pipeline. It was a lot of fun working on it and watching the HW engineers play around with it.

I believe Charles Irby would not be offended by getting rid of the mouse.

**Reply by harzner** [reply]
Wow, you surely have an impressive background, working with these innovative people and companies. Thank you for the insight. My idea has yet to prove itself, and I'm not sure wether it will or not, but I'd like to give it a try, and see what comes out of it. I will need some more time though to get back into coding.

And, I don't want to get rid of the mouse as such, teehee. I'm just asking myself wether I or me and anyone who's interested can create a development environment, where developers would not have to rely on it so much.

Gimme some time to restart my prototype. Once it runs, I'm surely gonna post about it here.

Kind regards,
Harzner

**Comment by harzner** [comment]
Thank you @VolkerEnderlein for your help. My little project still runs into problems linking the DLLs, but I'm quite motivated to solve that myself. I understand that CMake is the way to go, so I'm reading up on that, and learning how to use CMake.

Anyways, your reply helped me quite a lot, and pointed me into the right direction. Thank you again!

Kind regards,
Harzner

**Reply by VolkerEnderlein** [reply]
Hello Harzner,
thanks for providing feedback. Feel free to ask, as I know that setting up a CMake project with Qt and other 3rd party software can be a challenging task.
Cheers, Volker
